gate-web/src/main/groovy/com/netflix/spinnaker/gate/services/NotificationService.groovy

```diff
@@ -41,4 +60,38 @@ class NotificationService {
   void deleteNotificationConfig(String type, String app) {
     front50Service.deleteNotificationConfig(type, app)
   }
+
+  ResponseEntity<String> processNotificationCallback(String source, RequestEntity<String> request) {
+    String contentType = request.getHeaders().getFirst("Content-Type")?.toLowerCase()
+
+    if (!contentType) {
+      throw new InvalidRequestException("No Content-Type header present in request. Unable to process notification callback.")
+    }
+
+    final MediaType mediaType = MediaType.parse(contentType)
+
+    // We use the "raw" OkHttpClient here instead of EchoService because retrofit messes up with the encoding
+    // of the body for the x-www-form-urlencoded content type, which is what Slack uses. This allows us to pass
+    // the original body unmodified along to echo.
+    Endpoint echoEndpoint = serviceConfiguration.getServiceEndpoint("echo")
+
+    Request.Builder builder = new Request.Builder()
+      .url(echoEndpoint.url + request.url.path)
+      .post(RequestBody.create(mediaType, request.body))
+
+    request.getHeaders().each { String name, List values ->
```

Comment:
Can we be confident here that no `X-SPINNAKER-*` headers are going to be propagated?

I believe the http filter should be taking them out ... I'd be weary of arbitrarily passing any header through the stack but maybe there is something specific we need on the slack callback?

Reply:
There is. Slack adds an `X-SLACK-SIGNATURE` header that needs to be checked for security. I could not come up with a generic way to allow specific headers other than a bunch of configuration to whitelist them, and at the same time I couldn't think of a security concern to avoid passing them through, so I thought it was OK.

As for the Spinnaker headers, I'll confirm whether they're being added (in which case I could perhaps use `allowAnonymous` here?), but I also didn't think their addition would be a problem?